pkg/placement/placement_test.go

```diff
@@ -0,0 +1,178 @@
+
+	cleanUpFn := func() {
+		testServer.Shutdown()
+	}
+
+	serverAddress := "127.0.0.1:" + strconv.Itoa(port)
+	return serverAddress, testServer, cleanUpFn
+}
+
+func newTestClient(serverAddress string) (*grpc.ClientConn, v1pb.Placement_ReportDaprStatusClient, error) {
+	conn, err := grpc.Dial(serverAddress, grpc.WithInsecure())
+	if err != nil {
+		return nil, nil, err
+	}
+	client := v1pb.NewPlacementClient(conn)
+	stream, err := client.ReportDaprStatus(context.Background())
+	if err != nil {
+		conn.Close()
+		return nil, nil, err
+	}
+
+	return conn, stream, nil
+}
+
+func TestMemberRegistration(t *testing.T) {
+	serverAddress, testServer, cleanup := newTestPlacementServer(testRaftServer)
+
+	t.Run("Connect server and disconnect it gracefully", func(t *testing.T) {
+		// arrange
+		conn, stream, err := newTestClient(serverAddress)
+		assert.NoError(t, err)
+
+		host := &v1pb.Host{
+			Name:     "127.0.0.1:50102",
+			Entities: []string{"DogActor", "CatActor"},
+			Id:       "testAppID",
+			Load:     1, // Not used yet
+			// Port is redundant because Name should include port number
+		}
+
+		// act
+		stream.Send(host)
+
+		// assert
+		select {
+		case memberChange := <-testServer.membershipCh:
+			assert.Equal(t, raft.MemberUpsert, memberChange.cmdType)
+			assert.Equal(t, host.Name, memberChange.host.Name)
+			assert.Equal(t, host.Id, memberChange.host.AppID)
+			assert.EqualValues(t, host.Entities, memberChange.host.Entities)
+			assert.Equal(t, 1, len(testServer.streamConns))
+
+		case <-time.After(100 * time.Millisecond):
+			assert.True(t, false, "no membership change")
+		}
+
+		// act
+		// Runtime needs to close stream gracefully which will let placement remove runtime host from hashing ring
+		// in the next flush time window.
+		stream.CloseSend()
+
+		// assert
+		select {
+		case memberChange := <-testServer.membershipCh:
+			assert.Equal(t, raft.MemberRemove, memberChange.cmdType)
+			assert.Equal(t, host.Name, memberChange.host.Name)
+
+		case <-time.After(100 * time.Millisecond):
+			require.True(t, false, "no membership change")
+		}
+
+		conn.Close()
+	})
+
+	t.Run("Connect server and disconnect it forcefully", func(t *testing.T) {
+		// arrange
+		conn, stream, err := newTestClient(serverAddress)
+		assert.NoError(t, err)
+
+		// act
+		host := &v1pb.Host{
+			Name:     "127.0.0.1:50103",
+			Entities: []string{"DogActor", "CatActor"},
+			Id:       "testAppID",
+			Load:     1, // Not used yet
+			// Port is redundant because Name should include port number
+		}
+		stream.Send(host)
+
+		// assert
+		select {
+		case memberChange := <-testServer.membershipCh:
+			assert.Equal(t, raft.MemberUpsert, memberChange.cmdType)
+			assert.Equal(t, host.Name, memberChange.host.Name)
+			assert.Equal(t, host.Id, memberChange.host.AppID)
+			assert.EqualValues(t, host.Entities, memberChange.host.Entities)
+			assert.Equal(t, 1, len(testServer.streamConns))
+
+		case <-time.After(100 * time.Millisecond):
```

Comment:
should this be after flushTimer?

Reply:
no this is not related to flushTimer. This test is testing for ReportDaprStatus method. we can put any arbitrary number but I choose 100 millisecond to wait until ReportDaprStatus api queued member change to membershipCh channel.

Reply:
I defined constant to clarify the timeout usage.